src/Functions/hasAnyAllTokens.cpp

```diff
@@ -70,17 +70,40 @@ void FunctionHasAnyAllTokens<HasTokensTraits>::setSearchTokens(const std::vector
 namespace
 {
 
+/// Functions accept string as input (will be tokenized) or array of strings (used as-is)
 /// Also accepts Array(Nothing) which is the type of Array([])
-bool isArrayOfStringType(const IDataType & type)
+bool isStringOrArrayOfStringType(const IDataType & type)
 {
+    const auto * string_type = checkAndGetDataType<DataTypeString>(&type);
+    if (string_type)
+        return true;
+
     const auto * array_type = checkAndGetDataType<DataTypeArray>(&type);
-    if (!array_type)
-        return false;
+    if (array_type)
+    {
+        const DataTypePtr & nested_type = array_type->getNestedType();
+        return isString(nested_type) || isFixedString(nested_type) || isNothing(nested_type);
+    }
 
-    const DataTypePtr & nested_type = array_type->getNestedType();
-    return isString(nested_type) || isFixedString(nested_type) || isNothing(nested_type);
+    return false;
 }
 
+
+void fillNeedlesFromString(Needles & needles_tmp, const StringRef & needle_str)
+{
+    auto token_extractor_ = DefaultTokenExtractor();
+
+    size_t cur = 0;
+    size_t token_start = 0;
+    size_t token_len = 0;
+    size_t length = needle_str.size;
+    size_t pos = 0;
+
+    while (cur < length && token_extractor_.nextInStringPadded(needle_str.data, length, &cur, &token_start, &token_len))
+    {
+        needles_tmp.emplace(StringRef(needle_str.data + token_start, token_len), pos++);
```

Comment:
Minor: We are trying to get away from `StringRef` as much as possible and use `std::string_view` instead.

In l. 104, you can write

```cpp
needles_tmp.emplace(std::string{needle_str.data + token_start, token_len}, pos++);
```

In l. 242, you can write

```cpp
fillNeedlesFromString(needles_tmp, col_needles_str->getDataAt(0).toView());
```